storage/remote/client.go

```diff
@@ -38,30 +39,54 @@ const maxErrMsgLen = 256
 
 var userAgent = fmt.Sprintf("Prometheus/%s", version.Version)
 
-// Client allows reading and writing from/to a remote HTTP endpoint.
-type Client struct {
+// client allows reading and writing from/to a remote HTTP endpoint.
+type client struct {
 	remoteName string // Used to differentiate clients in metrics.
 	url        *config_util.URL
 	client     *http.Client
 	timeout    time.Duration
+
+	readQueries prometheus.Gauge
 }
 
-// ClientConfig configures a Client.
+// ClientConfig configures a client.
 type ClientConfig struct {
 	URL              *config_util.URL
 	Timeout          model.Duration
 	HTTPClientConfig config_util.HTTPClientConfig
 }
 
-// NewClient creates a new Client.
-func NewClient(remoteName string, conf *ClientConfig) (*Client, error) {
+// TODO(bwplotka): Add streamed chunked remote read method as well (https://github.com/prometheus/prometheus/issues/5926).
+type ReadClient interface {
+	Read(ctx context.Context, query *prompb.Query) (*prompb.QueryResult, error)
+}
+
+// NewReadClient creates a new client for remote read.
+// Caller has to ensure there are no duplicated clients.
+func NewReadClient(readQueries prometheus.Gauge, name string, conf *ClientConfig) (ReadClient, error) {
```

Comment:
Should we pass in a registry instead?

Reply:
lemme see

Reply:
Ok, so no, in this case metric is what we want because we can create and delete clients for the whole life of the process and as you commented that we cannot use `Unregister` it's more explicit and nicer to register those on Storage level which is singleton.

Reply:
It still feels weird to be passing metrics around. What happens if we add more metrics such as latency around?

Reply:
Why? We use that pattern in many many places in different projects to allow sharing the same `GaugeVec`. If we will have more than one we can create `readClientStruct` like we do in many places in Prometheus as well (:

Reply:
A given metric should only usually exist in a single function and file, if you're passing it around that indicates that the factoring probably isn't correct.

Reply:
This is not usual then. (: It's part of bigger system. :thinking: Let's talk offline, I really feel strong opinion that this is better than global metric. Cannot really see alternative here.

@beorn7 any feedback on this?

Reply:
Yes, and its lack of usualness is my concern. Instantiators of a class shouldn't care about what metrics it defines.

Reply:
That is fair statement. OK, I think I know the solution, let's just not export this. It is not used anywhere, just in `Storage` type. We can also have no constructor as it's constructed just once. I believe this will not violate `Instantiators of a class shouldn't care about what metrics it defines.` as it will internal type, tied to storage, in the same some `seriesSet` might have tied `seriesSetIterator` in it. WDYT?

Reply:
Changed, so it is not exported anymore, thus tied to stoage.

Reply:
Still seems a bit weird, visibility doesn't really change things.

Reply:
I think you can pass around a registry and then either register a `Gauge` with the `remoteName` and `endpoint` labels as const labels (the classical approach) or use [WrapRegistererWith](https://pkg.go.dev/github.com/prometheus/client_golang@v1.5.0/prometheus?tab=doc#WrapRegistererWith) for the same effect.

Reply:
Well that's the problem. None of those solutions works because we can create many clients of same labels for lifetime of the process (configure remote read for server example.com, then remove then configure again).

Reply:
That would help indeed. But my question is.. why complicating things? We use the same pattern as I suggest in a dozen of places in Thanos and it works really well. What's wrong with it? (: 

Anyway, don't mind to change to AreadyRegisteredError just curious.

Reply:
If a you get a AreadyRegisteredError I'd consider that a programming error and if you get it on purpose, poor architecture as you're trying to re-use metrics rather than define the right one once.

Reply:
Done. Also talked offline https://matrix.to/#/!WaUKIfoqfiyWQhenET:matrix.org/$15869391701329748dlPKk:matrix.org?via=matrix.org

Looks like it looks not bad,  happy with that. Thanks for discussion @brian-brazil and @beorn7. PTAL (:

Reply:
As indicated above, I consider any use of AreadyRegisteredError to be incorrect.

Reply:
I have to reuse one... that's the whole point.

Reply:
Ok at this point ... we are blocked. So far @brian-brazil  you did not propose ANY working idea and all my 4 attempts (Unregister, passing gauge, passing gauge on a private struct, AlreadyExistsError different) you rejected. So... what we do know? Because I am 100% sure the version with global is wrong. 

What would you suggest then?

Reply:
They way I'd approach this is to have a statically defined metric, pass down just the label value down from the instantiator, and have the instance access the child it needed via that label value.

Reply:
Search for Publisher on https://www.robustperception.io/label-lookups-and-the-child for the basic pattern I'd use.

Reply:
Hah that's interesting as the child pattern is exactly what I did with `newReadClient(readQueries.WithLabels(..., ...), name, ...)`, just the parent metric was not global/static (package level), but from parent struct. (:

Reply:
Otherwise all of it is package level so not really good solution.

Reply:
Anyway, looks like there is no agreement here, so, lets do this, let's choose whatever now from the 5 attempts we had and I will start separate issue with some options how we can fix many package level metric definitions (and If we want to fix it) in one go (:  WDYT? I can even revert to package level even though I think it's a bad solution.

Reply:
I find the (package-)global variable the most problematic of all suggested ways. It's not only a style question, it just hard-breaks if you import the package multiple times with different import paths (as it is common in Go anyway, but now also a core concept of Go modules).

I see drawbacks of both the AlreadyExistsError pattern and the unregister approach, but both are not too bad IMHO.

Reply:
Thanks. I am opening a separate issue about it will summarize our findings.

Let's focus on iterating over chunks here :hugs:

Reply:
> as the child pattern is exactly what I did with

No, that's quite different as you're passing the metric rather than the label value. Metric objects should not pass API boundaries.

> I see drawbacks of both the AlreadyExistsError pattern and the unregister approach

I think that any approach that relies on known anti-patterns is not a good approach, and we should not be encouraging such things by putting them inside Prometheus itself. Users looking to use these are usually either instrumenting at the wrong level, or should be writing a custom collector instead.

> I can even revert to package level even though I think it's a bad solution.

It's what the code already did, and it's a fine pattern that won't cause us any problems.

Reply:
> It's what the code already did, and it's a fine pattern that won't cause us any problems.

It's an anti-pattern, and it's fundamentally broken in any non-trivial code.